# CAGE-seq
library(CAGEr)
library(BSgenome.Glycine.max.T2T)
library(ggplot2)
library(dplyr)

setwd("/home/yanghy/my_data/SangData/CAGE-seq/RX2024-06016-G/seqkit/result_new/05.cager/")

## 1. Creating a CAGEexp object
pathsToInputFiles <- list.files(
  "/home/yanghy/my_data/SangData/CAGE-seq/RX2024-06016-G/seqkit/result_new/05.cager/bam", 
  full.names = TRUE,
  pattern = "\\.bam$"
)

samples <- c("Root_1", "Root_2", "Shoot_1", "Shoot_2")

CAGEset_Gmax <- CAGEexp(
  genomeName = "BSgenome.Glycine.max.T2T",
  inputFiles = pathsToInputFiles,
  inputFilesType = "bam",
  sampleLabels = samples
)

ce <- getCTSS(CAGEset_Gmax)
CTSStagCountSE(ce)

### CTSS coordinates
Gly_max_coordinate <- CTSScoordinatesGR(ce)
write.table(
  Gly_max_coordinate, 
  file = "Gly_max_coordinate.txt", 
  sep = "\t", 
  row.names = TRUE,
  quote = FALSE,
  col.names = TRUE
)

### Expression data
Gly_max_expression <- CTSStagCountDF(ce)
write.table(
  Gly_max_expression, 
  file = "Gly_max_expression.txt", 
  sep = "\t", 
  row.names = TRUE,
  quote = FALSE,
  col.names = TRUE
)

### Export CTSS counts for each sample
sample_names <- c("Root_1", "Root_2", "Shoot_1", "Shoot_2")
for (i in seq_along(sample_names)) {
  count_data <- CTSStagCountGR(ce, i)
  file_name <- paste0("Gly_max_count_", sample_names[i], ".txt")
  write.table(
    count_data,
    file = file.path(file_name),
    sep = "\t",
    row.names = FALSE,
    quote = FALSE,
    col.names = TRUE
  )
}

## 2. Genome annotations
gff_path <- "/home/yanghy/my_data/SangData/CAGE-seq/RX2024-06016-G/seqkit/15-50/clean/T2T/bowtie2/Ref/Glymax_cage.gff"
Gly_max_anno <- import(gff_path)
Gly_max_anno$transcript_type <- "protein_coding"
Gly_max_anno$transcript_type <- Rle(Gly_max_anno$transcript_type)
Gly_max_anno$type <- Rle(Gly_max_anno$type)
Gly_max_anno$gene_name <- Rle(Gly_max_anno$gene_name)
Gly_max_anno <- Gly_max_anno[which(Gly_max_anno$type %in% c("exon", "gene", "transcript")), ]
Gly_max_anno_1 <- Gly_max_anno[, c(9, 14, 2)]
ce_anno <- annotateCTSS(ce, Gly_max_anno_1, upstream = 1000, downstream = 500)
colData(ce_anno)[, c("librarySizes", "promoter", "exon", "intron", "unknown")]

pdf( "Glymax_anno_1000.pdf", width = 6, height = 4, pointsize = 12)
plotAnnot(x = ce_anno, scope = "counts") + 
  ggplot2::theme_bw() + 
  ggplot2::theme(legend.title = element_blank()) +
  ggplot2::theme(panel.grid = element_blank())
dev.off()

## 3. Sample Correlation
pdf( "Gly_max_correlation.pdf", width = 10, height = 9, pointsize = 12)
corr.m <- plotCorrelation2(
  ce_anno, 
  samples = "all", 
  tagCountThreshold = 1, 
  applyThresholdBoth = FALSE,
  method = "pearson",
  digits = 3
)
dev.off()

## 4. Merging of replicates
ce_merge <- mergeSamples(
  ce_anno, 
  mergeIndex = c(1, 1, 2, 2),
  mergedSampleLabels = c("Root", "Shoot")
)
ce_merge <- annotateCTSS(ce_merge, Gly_max_anno_1)

CTSS_Root <- CTSStagCountGR(ce_merge, 1)
CTSS_Shoot <- CTSStagCountGR(ce_merge, 2)

## 5. Normalization
pdf( "sample_fenbu30.pdf", width = 6, height = 6, pointsize = 12)
plotReverseCumulatives(ce_merge, fitInRange = c(5, 10000), onePlot = TRUE)
dev.off()

ce_normal <- normalizeTagCount(
  ce_merge,
  method = "powerLaw",
  fitInRange = c(5, 10000),
  alpha = 1.18,
  T = 10*10^5 
)

CTSS_TPM <- CTSSnormalizedTpmDF(ce_normal)
CTSS_Root_TPM <- CTSSnormalizedTpmGR(ce_normal, 1)
CTSS_Shoot_TPM <- CTSSnormalizedTpmGR(ce_normal, 2)

## 6. CTSS clustering
TC <- clusterCTSS(
  ce_normal,
  threshold = 1,
  nrPassThreshold = 1,
  thresholdIsTpm = TRUE,
  method = "distclu",
  maxDist = 150,
  removeSingletons = TRUE,
  keepSingletonsAbove = 1
)

rootend_TC <- tagClustersGR(TC, sample = "Root")
plantshoot_TC <- tagClustersGR(TC, sample = "Shoot")

## 7. Promoter width
TC_width <- cumulativeCTSSdistribution(TC, clusters = "tagClusters", useMulticore = TRUE)
TC_width <- quantilePositions(TC_width, clusters = "tagClusters", qLow = 0.1, qUp = 0.9)

rootend_width_tag <- tagClustersGR(
  TC_width, 
  "Root",
  returnInterquantileWidth = TRUE,
  qLow = 0.1,
  qUp = 0.9
)

plantshoot_width_tag <- tagClustersGR(
  TC_width, 
  "Shoot",
  returnInterquantileWidth = TRUE,
  qLow = 0.1,
  qUp = 0.9
)

pdf("TC_width_quant.pdf", width = 6, height = 4, pointsize = 12)
plotInterquantileWidth(
  TC_width, 
  clusters = "tagClusters", 
  tpmThreshold = 1, 
  qLow = 0.1, 
  qUp = 0.9
) +
  ggplot2::theme_bw() + 
  ggplot2::theme(legend.title = element_blank()) +
  ggplot2::theme(panel.grid = element_blank())
dev.off()

### Track export for genome browsers
iqtrack <- exportToTrack(
  TC_width, 
  what = "tagClusters", 
  qLow = 0.1, 
  qUp = 0.9, 
  oneTrack = FALSE
)

rtracklayer::export.bed(iqtrack, "TC-width.bed")

